test/src/testPubSub.cpp

```diff
@@ -0,0 +1,267 @@
+#include <gtest/gtest.h>
+
+#include <condition_variable>
+#include <mutex>
+#include <deque>
+#include <vector>
+#include <tuple>
+#include <chrono>
+#include <random>
+#include <sys/socket.h>
+#include <poll.h>
+#include <unistd.h>
+#include <signal.h>
+
+using namespace uprotocol::utransport;
+using namespace uprotocol::v1;
+using namespace uprotocol::uri;
+using namespace uprotocol::uuid;
+using namespace uprotocol::client;
+
+// capture and queue onReceive() for verification by the sender
+
+template <bool INTERPROC> 
+class MessageCapture : public UListener {
+public:
+    typedef std::vector<std::uint8_t>    Data;
+private:
+    std::condition_variable cv;
+    std::mutex mtx;
+    std::deque<Data>    queue;
+    int sp[2];
+public:
+    MessageCapture()
+    {
+        if constexpr (INTERPROC) {
+            socketpair(AF_UNIX, SOCK_DGRAM, 0, sp);
+        }
+    }
+
+    ~MessageCapture()
+    {
+        if constexpr (INTERPROC) {
+            close(sp[0]);
+            close(sp[1]);
+        }
+    }
+
+    void operator()(UMessage& message)
+    {
+        using namespace std;
+
+        if constexpr (INTERPROC) {
+            send(sp[0], message.payload().data(), message.payload().size(), 0);
+        }
+        else {
+            Data data(message.payload().size());
+            auto ptr = message.payload().data();
+            copy(ptr, ptr+message.payload().size(), data.data());
+            { 
+                lock_guard<mutex> lock(mtx);
+                queue.emplace_back(data);
+                cv.notify_one();
+            }
+        }
+    }
+
+    UStatus onReceive(UMessage& m) const override
+    {
+        using namespace std;
+        // scrape the const-ness off this thing
+        const_cast<MessageCapture*>(this)->operator()(m);
+        UStatus status;
+        status.set_code(UCode::OK);
+        return status;
+    }
+
+    std::tuple<size_t, Data, std::chrono::time_point<std::chrono::steady_clock>> get(size_t milliseconds_max = 1000)
+    {
+        using namespace std;
+        using namespace std::chrono;
+
+        auto end_time = steady_clock::now() + milliseconds(milliseconds_max);
+        if constexpr (INTERPROC) {
+            Data data(2048);
+            struct pollfd fds[1];
+            fds[0].fd = sp[1];
+            fds[0].events = POLLIN;
+            fds[0].revents = 0;
+            auto ret = poll(fds, 1, milliseconds_max);
+            if (ret <= 0) return make_tuple(0, Data(), end_time);
+            ret = recv(sp[1], data.data(), 2048, 0);
+            if (ret <= 0) return make_tuple(0, Data(), end_time);
+            end_time = steady_clock::now();
+            data.resize(ret);
+            return make_tuple(1, data, end_time);
+        }
+        else {
+            size_t count;
+            Data data;
+            {
+                unique_lock<mutex> lock(mtx);
+                while ((count = queue.size()) < 1) cv.wait_until(lock, end_time);
```

Comment:
Minor: predicate form of `wait_until` can be used here
```suggestion
                cv.wait_until(lock, end_time, [&count, &queue](){ count = queue.size(); return count >= 1; });
```

Reply:
Adopting this in next push.